ui/src/SoundWave.js

```diff
@@ -0,0 +1,70 @@
+export class SoundWave {
+
+    constructor(canvas, stream) {
+        this.canvas = canvas;
+        this.stream = stream;
+        this.analyser = undefined;
+        this.stopId = undefined;
+        this.freqs = undefined;
+        this.ctx = undefined;
+        this.draw = this.draw.bind(this);
```

Comment:
`this` should already be bound to `this.draw`, it's a class method. Are you getting `this` undefined errors without the bind?

Reply:
Yes. I get an error saying `Cannot read property 'ctx' of undefined at draw` without it

Reply:
It should work if you pass an arrow function which does not bind a new this, into requestAnmitationFrame

```javascript
requestAnimationFrame(() => this.draw());
```